flink-ml/flink-ml-api/src/test/java/org/apache/flink/ml/api/core/PipelineTest.java

```diff
@@ -0,0 +1,181 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.api.core;
+
+import org.apache.flink.ml.api.misc.param.ParamInfo;
+import org.apache.flink.ml.api.misc.param.ParamInfoFactory;
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.ml.util.persist.PipelineStageFactory;
+import org.apache.flink.table.api.Table;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/**
+ * Tests for pipeline behavior.
+ */
+public class PipelineTest {
+	@Rule
+	public ExpectedException thrown = ExpectedException.none();
+
+	@Test
+	public void testPipelineBehavior() {
+		Pipeline pipeline = new Pipeline();
+		pipeline.appendStage(new MockTransformer("a"));
+		pipeline.appendStage(new MockEstimator("b"));
+		pipeline.appendStage(new MockEstimator("c"));
+		pipeline.appendStage(new MockTransformer("d"));
+		assert describePipeline(pipeline).equals("a_b_c_d");
+
+		Pipeline pipelineModel = pipeline.fit(null);
+		assert describePipeline(pipelineModel).equals("a_mb_mc_d");
+
+		thrown.expect(RuntimeException.class);
+		thrown.expectMessage("Pipeline contains Estimator, need to fit first.");
+		pipeline.transform(null);
+	}
+
+	@Test
+	public void testPipelineRestore() {
+		Pipeline pipeline = new Pipeline();
+		pipeline.appendStage(new MockTransformer("a"));
+		pipeline.appendStage(new MockEstimator("b"));
+		pipeline.appendStage(new MockEstimator("c"));
+		pipeline.appendStage(new MockTransformer("d"));
+		String pipelineJson = pipeline.toJson();
+
+		Pipeline restoredPipeline = PipelineStageFactory.createFromJson(pipelineJson);
+		assert describePipeline(restoredPipeline).equals("a_b_c_d");
+
+		Pipeline pipelineModel = pipeline.fit(null);
+		String modelJson = pipelineModel.toJson();
+
+		Pipeline restoredPipelineModel = PipelineStageFactory.createFromJson(modelJson);
+		assert describePipeline(restoredPipelineModel).equals("a_mb_mc_d");
+	}
+
+	private static String describePipeline(Pipeline p) {
+		StringBuilder res = new StringBuilder();
+		for (PipelineStage s : p.getStages()) {
+			if (res.length() != 0) {
+				res.append("_");
+			}
+			res.append(((SelfDescribe) s).describe());
+		}
+		return res.toString();
+	}
+
+	/**
+	 * Interface to describe a class with a string, only for pipeline test.
+	 */
+	public interface SelfDescribe {
```

Comment:
I think, that all implementations of this interface are stubs - not mock, because you have implemented some logic for check behavior. I may be mistaken, please correct me

Reply:
I'm not sure about the difference between stub and mock. Let me find it out first.

Other advices seem reasonable. I'll take them. Thanks.